src/System.Collections/src/System/Collections/Generic/Queue.cs

```diff
@@ -252,16 +255,19 @@ public T Dequeue()
 
         public bool TryDequeue(out T result)
         {
-            if (_size == 0)
+            int head = _head;
+            T[] array = _array;
+
+            if (_size == 0 || (uint)head >= (uint)array.Length)
```

Comment:
Same question; why is the latter condition necessary?

Reply:
Same reason as above.

> To enable the range check elimination on the following `array[head]`. The real win is on reference types, where this access is two times.

Reply:
I'm investigating this case right now.

Reply:
No clear difference measurable. The asm is quite similar, except the jumps at the beginning of the method.

With manual range check:
```asm
G_M54913_IG02:
       448B7318             mov      r14d, dword ptr [rbx+24]
       4C8B7B08             mov      r15, gword ptr [rbx+8]
       837B2000             cmp      dword ptr [rbx+32], 0
       7409                 je       SHORT G_M54913_IG03
       418B4708             mov      eax, dword ptr [r15+8]
       413BC6               cmp      eax, r14d
       7710                 ja       SHORT G_M54913_IG05

G_M54913_IG03:
       33C0                 xor      rax, rax
       488906               mov      qword ptr [rsi], rax

G_M54913_IG04:
       488D65E8             lea      rsp, [rbp-18H]
       5B                   pop      rbx
       415E                 pop      r14
       415F                 pop      r15
       5D                   pop      rbp
       C3                   ret      

G_M54913_IG05:
       488975D8             mov      bword ptr [rbp-28H], rsi
       4963FE               movsxd   rdi, r14d
       498B74FF10           mov      rsi, gword ptr [r15+8*rdi+16]
       488B7DD8             mov      rdi, bword ptr [rbp-28H]
       E8CF54AD78           call     CORINFO_HELP_CHECKED_ASSIGN_REF
      
; more code (equal)
```

Without manual range check:
```asm
G_M50722_IG02:
       448B7318             mov      r14d, dword ptr [rbx+24]
       4C8B7B08             mov      r15, gword ptr [rbx+8]
       837B2000             cmp      dword ptr [rbx+32], 0
       7510                 jne      SHORT G_M50722_IG04
       33C0                 xor      rax, rax
       488906               mov      qword ptr [rsi], rax

G_M50722_IG03:
       488D65E8             lea      rsp, [rbp-18H]
       5B                   pop      rbx
       415E                 pop      r14
       415F                 pop      r15
       5D                   pop      rbp
       C3                   ret      

G_M50722_IG04:
       488975D8             mov      bword ptr [rbp-28H], rsi
       418B7F08             mov      edi, dword ptr [r15+8]
       443BF7               cmp      r14d, edi
       7374                 jae      SHORT G_M50722_IG09
       4963FE               movsxd   rdi, r14d
       498B74FF10           mov      rsi, gword ptr [r15+8*rdi+16]
       488B7DD8             mov      rdi, bword ptr [rbp-28H]
       E8EF53AD78           call     CORINFO_HELP_CHECKED_ASSIGN_REF

; more code (equal)

G_M50722_IG09:
       E896B79678           call     CORINFO_HELP_RNGCHKFAIL
       CC                   int3   
```

So I'll also go with the cleaner code and remove the manual range check?

Reply:
> So I'll also go with the cleaner code and remove the manual range check?

Yes please.  Thanks.